servers/physics_3d/godot_body_pair_3d.cpp

```diff
@@ -591,7 +591,7 @@ void GodotBodySoftBodyPair3D::contact_added_callback(const Vector3 &p_point_A, i
 	contact.index_B = p_index_B;
 	contact.local_A = local_A;
 	contact.local_B = local_B;
-	contact.normal = (p_point_A - p_point_B).normalized();
+	contact.normal = (normal.dot((p_point_A - p_point_B)) < 0 ? -normal : normal);
```

Comment:
The normal vector in the callbacks seems to always point towards A. For example _rest_cbk_result just trusts the normal is pointing in the correct direction. Why does this one have to do a direction check?

Reply:
I found cases that I couldn't figure out any other way to handle.  For example, in the `_generate_contacts_X_X()` functions, you may have three points that define a face or a circle, but you don't know which side is the inside.  Perhaps there's some way to figure out?  And then there's `_generate_contacts_edge_edge()`.  We know the normal should be perpendicular to the two edges, but that leaves the sign undefined.

Reply:
If that's the case, wouldn't you need to put the same logic in other callback functions:
 - `GodotBodyPair3D::_contact_added_callback`
 - `GodotCollisionSolver3D::soft_body_contact_callback`
 - `_rest_cbk_result`

Conversely you could correct the normal direction only for those cases where it is ambiguous (in the `_generate_contacts_X_X()` functions, and then the callback functions wouldn't need to handle the possibility of receiving direction-flipped normals.

Reply:
The individual callback functions don't require it.  It's handled by `_CollectorCallback.call()` before it invokes the callback.  (I know, it's confusing!  We have a "callback" object that invokes a "callback" function.)  Since everything goes through that one function, nothing else needs to worry about the direction, neither the individual contact detection functions nor the callback functions.

So why is it needed here?  Possibly because I made a mistake `_CollectorCallback`! 😀  It forces the normal to point from A to B rather than B to A, while this routine expects it to be the other way.  Though the reason I did it that way was that your PhysicsCollisionTest seemed to expect it, so I assumed that was the correct direction.  If I make the normal point from B to A, the bar indicating the normal disappears, or rather is mostly hidden as it points down into the large object.  But is that supposed to be the correct direction?

Reply:
@reduz I think this is a question for you.  Which way should the normals point, from A to B or from B to A?  Is it really intended that soft body collisions use normals that point one way, and all other collisions have them point the other way?